src/jit/smallhash.h

```diff
@@ -0,0 +1,605 @@
+        }
+
+        inline TValue& Value()
+        {
+            return m_bucket->m_value;
+        }
+    };
+
+    class Iterator final
+    {
+        friend class HashTableBase<TKey, TValue, TKeyInfo>;
+
+        Bucket* m_buckets;
+        unsigned m_numBuckets;
+        unsigned m_index;
+
+        Iterator(Bucket* buckets, unsigned numBuckets, unsigned index)
+            : m_buckets(buckets)
+            , m_numBuckets(numBuckets)
+            , m_index(index)
+        {
+            assert((buckets != nullptr) || (numBuckets == 0));
+            assert(index <= numBuckets);
+
+            // Advance to the first occupied bucket
+            while (m_index != m_numBuckets && !m_buckets[m_index].m_isFull)
+            {
+                m_index++;
+            }
+        }
+
+    public:
+        Iterator()
+            : m_buckets(nullptr)
+            , m_numBuckets(0)
+            , m_index(0)
+        {
+        }
+
+        KeyValuePair operator*() const
+        {
+            if (m_index >= m_numBuckets)
+            {
+                return KeyValuePair();
+            }
+
+            Bucket* bucket = &m_buckets[m_index];
+            assert(bucket->m_isFull);
+            return KeyValuePair(bucket);
+        }
+
+        KeyValuePair operator->() const
+        {
+            return this->operator*();
+        }
+
+        bool operator==(const Iterator& other) const
+        {
+            return (m_buckets == other.m_buckets) && (m_index == other.m_index);
+        }
+
+        bool operator!=(const Iterator& other) const
+        {
+            return (m_buckets != other.m_buckets) || (m_index != other.m_index);
+        }
+
+        Iterator& operator++()
+        {
+            do
+            {
+                m_index++;
+            } while (m_index != m_numBuckets && !m_buckets[m_index].m_isFull);
+
+            return *this;
+        }
+    };
+
+    Iterator begin() const
+    {
+        return Iterator(m_buckets, m_numBuckets, 0);
+    }
+
+    Iterator end() const
+    {
+        return Iterator(m_buckets, m_numBuckets, m_numBuckets);
+    }
+
+    unsigned Count() const
+    {
+        return m_numFullBuckets;
+    }
+
+    void Clear()
+    {
+        if (m_numBuckets > 0)
+        {
+            memset(m_buckets, 0, sizeof(Bucket) * m_numBuckets);
+            m_numFullBuckets = 0;
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // HashTableBase::AddOrUpdate: adds a key-value pair to the hash table if
+    //                             the key does not already exist in the
+    //                             table, or updates the value if the key
+    //                             already exists.
+    //
+    // Arguments:
+    //    key   - The key for which to add or update a value.
+    //    value - The value.
+    //
+    // Returns:
+    //    True if the value was added; false if it was updated.
+    bool AddOrUpdate(TKey& key, TValue&& value)
```

Comment:
> Shouldn't key be const TKey&?

Yes. Fixed.

> Also, I think key being a reference is debatable in the context of a compiler. Usually references are used in case key copying is potentially expensive, like it happens for strings. But it's unlikely that in a compiler the key will ever be something other than a pointer or an integral type so using a reference is an unnecessary expense.
> 
> A good compromise might be to have the parameter type be provide by HashTableInfo via a typedef.

This is an interesting idea. It adds a bit of complication, but could ensure that we get the code gen we want. Do you know if MSVC/Clang will optimize a `const T&` parameter into a by-value parameter if it's not address-exposed?